File: src/main/java/com/epri/dss/general/impl/XYCurveObjImpl.java
package com.epri.dss.general.impl;

import java.io.PrintStream;

import com.epri.dss.common.DSSClass;
import com.epri.dss.general.XYCurve;
import com.epri.dss.general.XYCurveObj;

public class XYCurveObjImpl extends DSSObjectImpl implements XYCurveObj {

	private int LastValueAccessed,
			NumPoints;  // number of points in curve
	private int ArrayPropertyIndex;
	private double X, Y;
	private double[] XValues, YValues;

	public XYCurveObjImpl(DSSClass ParClass, String XYCurveName) {
		super(ParClass);
		setName(XYCurveName.toLowerCase());
		this.objType = ParClass.getDSSClassType();

		this.LastValueAccessed = 0;

		this.NumPoints = 0;
		this.XValues = null;
		this.YValues = null;

		this.X = 0.0;
		this.Y = 0.0;

		ArrayPropertyIndex = -1;

		initPropertyValues(0);
	}

	/**
	 * Get Y value at specified X value.
	 *
	 * This method returns the interpolated Y value for the given X.
	 * If no points exist in the curve, the result is  0.0
	 * If Xvalue is outside the range of defined X values,
	 * the curve is extrapolated from the ends.
	 */
	public double getYValue(double X) {
		double Result = 0.0;  // default return value if no points in curve

		if (NumPoints > 0)  // handle exceptional cases
			if (NumPoints == 1) {
				return YValues[0];
			} else {
				/* Start with previous value accessed under the assumption that most
				 * of the time, the values won't change much.
				 */

				if (XValues[LastValueAccessed] > X)
					LastValueAccessed = 1;  // start over from beginning

				// if off the curve for the first point, extrapolate from the first two points
				if ((LastValueAccessed == 0) && (XValues[0] > X)) {
					Result = interpolatePoints(0, 1, X, XValues, YValues);
					return Result;
				}

				// in the middle of the arrays
				for (int i = LastValueAccessed; i < NumPoints; i++) {
					if (Math.abs(XValues[i] - X) < 0.00001) {  // if close to an actual point, just use it
						Result = YValues[i];
						LastValueAccessed = i;
						return Result;
					} else if (XValues[i] > X) {
						// interpolate between two values
						LastValueAccessed = i - 1;  // TODO Check zero based indexing
						Result = interpolatePoints(i, LastValueAccessed, X, XValues, YValues);
						return Result;
					}
				}

				// if we fall through the loop, Extrapolate from last two points
				LastValueAccessed = NumPoints - 1;
				Result = interpolatePoints(NumPoints, LastValueAccessed,  X, XValues, YValues);
			}

		return Result;
	}

	/**
	 * Get Y value by index.
	 */
	public double getYValue(int i) {
		if ((i < NumPoints) && (i >= 0)) {
			LastValueAccessed = i;
			return YValues[i];
		} else {
			return 0.0;
		}
	}

	/**
	 * Get X value corresponding to point index.
	 */
	public double getXValue(int i) {
		if ((i < NumPoints) && (i >= 0)) {
			LastValueAccessed = i;
			return XValues[i];
		} else {
			return 0.0;
		}
	}

	@Override
	public void dumpProperties(PrintStream F, boolean Complete) {
		super.dumpProperties(F, Complete);

		for (int i = 0; i < getParentClass().getNumProperties(); i++) {
			switch (i) {
			case 2:
				F.println("~ " + getParentClass().getPropertyName()[i] + "=(" + getPropertyValue(i) + ")");
				break;
			case 3:
				F.println("~ " + getParentClass().getPropertyName()[i] + "=(" + getPropertyValue(i) + ")");
				break;
			default:
				F.println("~ " + getParentClass().getPropertyName()[i] + "=" + getPropertyValue(i));
				break;
			}
		}
	}

	@Override
	public String getPropertyValue(int Index) {
		String Result;

		switch (Index) {
		case 1:
			Result = "[";
			break;
		case 2:
			Result = "[";
			break;
		case 3:
			Result = "[";
			break;
		default:
			Result = "";
			break;
		}

		switch (Index) {
		case 1:
			if ((XValues != null) && (YValues != null)) {
				for (int i = 0; i < NumPoints; i++)
					Result = Result + String.format("%.8g, %.8g ", XValues[i], YValues[i]);
			} else {
				Result = "0, 0";
			}
			break;
		case 2:
			if (YValues != null) {
				for (int i = 0; i < NumPoints; i++)
					Result = Result + String.format("%-g, ", YValues[i]);
			} else {
				Result = "0";
			}
			break;
		case 3:
			if (XValues != null) {
				for (int i = 0; i < NumPoints; i++)
					Result = Result + String.format("%-g, ", XValues[i]);
			} else {
				Result = "0";
			}
			break;
		case 7:
			Result = String.format("%.8g", getXValue(Y));
			break;
		case 8:
			Result = String.format("%.8g", getYValue(X));
			break;
		default:
			Result = super.getPropertyValue(Index);
			break;
		}

		switch (Index) {
		case 1:
			Result = "[";
			break;
		case 2:
			Result = "[";
			break;
		case 3:
			Result = "[";
			break;
		}

		return Result;
	}

	/**
	 * Get X value at specified Y value.
	 *
	 * This method returns the interpolated X value for the given Y.
	 * If no points exist in the curve, the result is 0.0
	 * If Xvalue is outside the range of defined X values,
	 * the curve is extrapolated from the Ends.
	 */
	public double getXValue(double Y) {
		double Result = 0.0;  // default return value if no points in curve

		if (NumPoints > 0)  // handle exceptional cases
			if (NumPoints == 1) {
				Result = XValues[0];
			} else {
				/* Start with previous value accessed under the assumption that most
				 * of the time, this function will be called sequentially
				 */

				if (YValues[LastValueAccessed] > Y)
					LastValueAccessed = 0;  // start over from beginning

				// if off the curve for the first point, extrapolate from the first two points
				if ((LastValueAccessed == 0) && (YValues[0] > Y)) {
					Result = interpolatePoints(0, 1, Y, YValues, XValues);
					return Result;
				}

				for (int i = LastValueAccessed; i < NumPoints; i++) {  // TODO Check zero based indexing
					if (Math.abs(YValues[i] - Y) < 0.00001) {  // if close to an actual point, just use it.
						Result = XValues[i];
						LastValueAccessed = i;
						return Result;
					} else if (YValues[i] > Y) {
						// interpolate
						LastValueAccessed = i - 1;  // TODO Check zero based indexing
						Result = interpolatePoints(i, LastValueAccessed, Y, YValues, XValues);
						return Result;
					}
				}

				// if we fall through the loop, extrapolate from last two points
				LastValueAccessed = NumPoints - 1;
				Result = interpolatePoints(NumPoints, LastValueAccessed,  Y, YValues, XValues);
			}

		return Result;
	}

	@Override
	public void initPropertyValues(int ArrayOffset) {

		setPropertyValue(0, "0");  // number of points to expect
		setPropertyValue(1, "");
		setPropertyValue(2, "");
		setPropertyValue(3, "");
		setPropertyValue(4, "");
		setPropertyValue(5, "");
		setPropertyValue(6, "");

		super.initPropertyValues(XYCurve.NumPropsThisClass);
	}

	private double interpolatePoints(int i, int j, double X, double[] Xarray, double[] Yarray) {
		double Den = (Xarray[i] - Xarray[j]);
		if (Den != 0.0) {
			return Yarray[j] + (X - Xarray[j]) / Den * (Yarray[i] - Yarray[j]);
		} else {
			return Yarray[i];  // Y is undefined, return i-th value
		}
	}

	public int getNumPoints() {
		return NumPoints;
	}

	public void setNumPoints(int numPoints) {
		setPropertyValue(0, String.valueOf(numPoints));  // update property list variable

		// reset array property values to keep them in proper order in save

		if (ArrayPropertyIndex >= 0)
			setPropertyValue(ArrayPropertyIndex, getPropertyValue(ArrayPropertyIndex));

		NumPoints = numPoints;  // now assign the value
	}

	// FIXME Private members in OpenDSS

	public int getLastValueAccessed() {
		return LastValueAccessed;
	}

	public void setLastValueAccessed(int lastValueAccessed) {
		LastValueAccessed = lastValueAccessed;
	}

	public int getArrayPropertyIndex() {
		return ArrayPropertyIndex;
	}

	public void setArrayPropertyIndex(int arrayPropertyIndex) {
		ArrayPropertyIndex = arrayPropertyIndex;
	}

	public double getX() {
		return X;
	}

	public void setX(double x) {
		X = x;
	}

	public double getY() {
		return Y;
	}

	public void setY(double y) {
		Y = y;
	}

	public double[] getXValues() {
		return XValues;
	}

	public void setXValues(double[] xValues) {
		XValues = xValues;
	}

	public double[] getYValues() {
		return YValues;
	}

	public void setYValues(double[] yValues) {
		YValues = yValues;
	}

}

File: pydss/general/XYCurveObjImpl.py
from __pyjamas__ import (ARGERROR,)
from dss.general.impl.DSSObjectImpl import (DSSObjectImpl,)
from dss.general.XYCurveObj import (XYCurveObj,)
from dss.general.XYCurve import (XYCurve,)


class XYCurveObjImpl(DSSObjectImpl, XYCurveObj):
    lastValueAccessed = None
    numPoints = None
    # number of points in curve
    arrayPropertyIndex = None
    X = None
    Y = None
    XValues = None
    YValues = None

    def __init__(self, parClass, XYCurveName):
        super(XYCurveObjImpl, self)(parClass)
        self.setName(XYCurveName.toLowerCase())
        self.objType = parClass.getDSSClassType()
        self.lastValueAccessed = 0
        self.numPoints = 0
        self.XValues = None
        self.YValues = None
        self.X = 0.0
        self.Y = 0.0
        self.arrayPropertyIndex = -1
        self.initPropertyValues(0)

    def getYValue(self, *args):
        """Get Y value at specified X value.

        This method returns the interpolated Y value for the given X.
        If no points exist in the curve, the result is  0.0
        If Xvalue is outside the range of defined X values,
        the curve is extrapolated from the ends.
        ---
        Get Y value by index.
        """
        _0 = args
        _1 = len(args)
        if _1 == 1:
            if isinstance(_0[0], double):
                X, = _0
                Result = 0.0
                # default return value if no points in curve
                if self.numPoints > 0:
                    # handle exceptional cases
                    if self.numPoints == 1:
                        return self.YValues[0]
                    else:
                        # Start with previous value accessed under the assumption that most
                        # of the time, the values won't change much.

                        if self.XValues[self.lastValueAccessed] > X:
                            self.lastValueAccessed = 1
                            # start over from beginning
                            # if off the curve for the first point, extrapolate from the first two points
                        if self.lastValueAccessed == 0 and self.XValues[0] > X:
                            Result = self.interpolatePoints(0, 1, X, self.XValues, self.YValues)
                            return Result
                        # in the middle of the arrays
                        _0 = True
                        i = self.lastValueAccessed
                        while True:
                            if _0 is True:
                                _0 = False
                            else:
                                i += 1
                            if not (i < self.numPoints):
                                break
                            if self.Math.abs(self.XValues[i] - X) < 1.0000000000000002e-05:
                                # if close to an actual point, just use it
                                Result = self.YValues[i]
                                self.lastValueAccessed = i
                                return Result
                            elif self.XValues[i] > X:
                                # interpolate between two values
                                self.lastValueAccessed = i - 1
                                # TODO Check zero based indexing
                                Result = self.interpolatePoints(i, self.lastValueAccessed, X, self.XValues, self.YValues)
                                return Result
                        # if we fall through the loop, Extrapolate from last two points
                        self.lastValueAccessed = self.numPoints - 1
                        Result = self.interpolatePoints(self.numPoints, self.lastValueAccessed, X, self.XValues, self.YValues)
                return Result
            else:
                i, = _0
                if i < self.numPoints and i >= 0:
                    self.lastValueAccessed = i
                    return self.YValues[i]
                else:
                    return 0.0
        else:
            raise ARGERROR(1, 1)

    def getXValue(self, *args):
        """Get X value corresponding to point index.
        ---
        Get X value at specified Y value.

        This method returns the interpolated X value for the given Y.
        If no points exist in the curve, the result is 0.0
        If Xvalue is outside the range of defined X values,
        the curve is extrapolated from the Ends.
        """
        _0 = args
        _1 = len(args)
        if _1 == 1:
            if isinstance(_0[0], double):
                Y, = _0
                result = 0.0
                # default return value if no points in curve
                if self.numPoints > 0:
                    # handle exceptional cases
                    if self.numPoints == 1:
                        result = self.XValues[0]
                    else:
                        # Start with previous value accessed under the assumption that most
                        # of the time, this function will be called sequentially

                        if self.YValues[self.lastValueAccessed] > Y:
                            self.lastValueAccessed = 0
                            # start over from beginning
                            # if off the curve for the first point, extrapolate from the first two points
                        if self.lastValueAccessed == 0 and self.YValues[0] > Y:
                            result = self.interpolatePoints(0, 1, Y, self.YValues, self.XValues)
                            return result
                        _0 = True
                        i = self.lastValueAccessed
                        while True:
                            if _0 is True:
                                _0 = False
                            else:
                                i += 1
                            if not (i < self.numPoints):
                                break
                            # TODO Check zero based indexing
                            if self.Math.abs(self.YValues[i] - Y) < 1.0000000000000002e-05:
                                # if close to an actual point, just use it.
                                result = self.XValues[i]
                                self.lastValueAccessed = i
                                return result
                            elif self.YValues[i] > Y:
                                # interpolate
                                self.lastValueAccessed = i - 1
                                # TODO Check zero based indexing
                                result = self.interpolatePoints(i, self.lastValueAccessed, Y, self.YValues, self.XValues)
                                return result
                        # if we fall through the loop, extrapolate from last two points
                        self.lastValueAccessed = self.numPoints - 1
                        result = self.interpolatePoints(self.numPoints, self.lastValueAccessed, Y, self.YValues, self.XValues)
                return result
            else:
                i, = _0
                if i < self.numPoints and i >= 0:
                    self.lastValueAccessed = i
                    return self.XValues[i]
                else:
                    return 0.0
        else:
            raise ARGERROR(1, 1)

    def dumpProperties(self, f, complete):
        super(XYCurveObjImpl, self).dumpProperties(f, complete)
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.getParentClass().getNumProperties()):
                break
            _1 = i
            _2 = False
            while True:
                if _1 == 2:
                    _2 = True
                    print '~ ' + self.getParentClass().getPropertyName()[i] + '=(' + self.getPropertyValue(i) + ')'
                    break
                if (_2 is True) or (_1 == 3):
                    _2 = True
                    print '~ ' + self.getParentClass().getPropertyName()[i] + '=(' + self.getPropertyValue(i) + ')'
                    break
                if True:
                    _2 = True
                    print '~ ' + self.getParentClass().getPropertyName()[i] + '=' + self.getPropertyValue(i)
                    break
                break

    def getPropertyValue(self, index):
        _0 = index
        _1 = False
        while True:
            if _0 == 1:
                _1 = True
                result = '['
                break
            if (_1 is True) or (_0 == 2):
                _1 = True
                result = '['
                break
            if (_1 is True) or (_0 == 3):
                _1 = True
                result = '['
                break
            if True:
                _1 = True
                result = ''
                break
            break
        _2 = index
        _3 = False
        while True:
            if _2 == 1:
                _3 = True
                if self.XValues is not None and self.YValues is not None:
                    _4 = True
                    i = 0
                    while True:
                        if _4 is True:
                            _4 = False
                        else:
                            i += 1
                        if not (i < self.numPoints):
                            break
                        result = result + String.format.format('%.8g, %.8g ', self.XValues[i], self.YValues[i])
                else:
                    result = '0, 0'
                break
            if (_3 is True) or (_2 == 2):
                _3 = True
                if self.YValues is not None:
                    _5 = True
                    i = 0
                    while True:
                        if _5 is True:
                            _5 = False
                        else:
                            i += 1
                        if not (i < self.numPoints):
                            break
                        result = result + String.format.format('%-g, ', self.YValues[i])
                else:
                    result = '0'
                break
            if (_3 is True) or (_2 == 3):
                _3 = True
                if self.XValues is not None:
                    _6 = True
                    i = 0
                    while True:
                        if _6 is True:
                            _6 = False
                        else:
                            i += 1
                        if not (i < self.numPoints):
                            break
                        result = result + String.format.format('%-g, ', self.XValues[i])
                else:
                    result = '0'
                break
            if (_3 is True) or (_2 == 7):
                _3 = True
                result = String.format.format('%.8g', self.getXValue(self.Y))
                break
            if (_3 is True) or (_2 == 8):
                _3 = True
                result = String.format.format('%.8g', self.getYValue(self.X))
                break
            if True:
                _3 = True
                result = super(XYCurveObjImpl, self).getPropertyValue(index)
                break
            break
        _7 = index
        _8 = False
        while True:
            if _7 == 1:
                _8 = True
                result = '['
                break
            if (_8 is True) or (_7 == 2):
                _8 = True
                result = '['
                break
            if (_8 is True) or (_7 == 3):
                _8 = True
                result = '['
                break
            break
        return result

    def initPropertyValues(self, arrayOffset):
        self.setPropertyValue(0, '0')
        # number of points to expect
        self.setPropertyValue(1, '')
        self.setPropertyValue(2, '')
        self.setPropertyValue(3, '')
        self.setPropertyValue(4, '')
        self.setPropertyValue(5, '')
        self.setPropertyValue(6, '')
        super(XYCurveObjImpl, self).initPropertyValues(XYCurve.NumPropsThisClass)

    def interpolatePoints(self, i, j, X, XArray, YArray):
        den = XArray[i] - XArray[j]
        if den != 0.0:
            return YArray[j] + (((X - XArray[j]) / den) * (YArray[i] - YArray[j]))
        else:
            return YArray[i]
            # Y is undefined, return i-th value

    def getNumPoints(self):
        return self.numPoints

    def setNumPoints(self, num):
        # FIXME Private members in OpenDSS
        self.setPropertyValue(0, String.valueOf.valueOf(num))
        # update property list variable
        # reset array property values to keep them in proper order in save
        if self.arrayPropertyIndex >= 0:
            self.setPropertyValue(self.arrayPropertyIndex, self.getPropertyValue(self.arrayPropertyIndex))
        self.numPoints = num
        # now assign the value

    def getLastValueAccessed(self):
        return self.lastValueAccessed

    def setLastValueAccessed(self, lastValue):
        self.lastValueAccessed = lastValue

    def getArrayPropertyIndex(self):
        return self.arrayPropertyIndex

    def setArrayPropertyIndex(self, index):
        self.arrayPropertyIndex = index

    def getX(self):
        return self.X

    def setX(self, x):
        self.X = x

    def getY(self):
        return self.Y

    def setY(self, y):
        self.Y = y

    def getXValues(self):
        return self.XValues

    def setXValues(self, xvalues):
        self.XValues = xvalues

    def getYValues(self):
        return self.YValues

    def setYValues(self, yvalues):
        self.YValues = yvalues
